### ISSUE TITLE
Puzzling frozen array for unloadable_cpaths.

### ISSUE THREAD

**Issue by pat** [fxn/zeitwerk#56]
I'm not sure if this should be logged here or in ActiveSupport, given it's overlapping between the two… but also, I'm not entirely across the interplay of all of this anyway.

I've noticed that the value for `unloadable_cpaths` is a frozen array. This is passed through to Rails as the initial `autoloaded_constants` value in `ActiveSupport::Dependencies` - but the logic there modifies the provided array (with `uniq!` and `concat` calls in particular). At a high level, I'm finding this confusing - why provide a frozen object for something where modifications are expected?

On a more tangible/practical level: Thinking Sphinx (a gem I maintain) loads all files within `app/indices` (to find index definitions) using `ActiveSupport::Dependencies.require_or_load file`. When this occurs in Rails 6.0.0.rc1, a `FrozenError` crops up, because the underlying modifications to `autoloaded_constants` are being attempted on that initial frozen value (as noted in pat/thinking-sphinx#1137).

So: is this a bug in Zeitwerk/ActiveSupport? Or am I misusing/misunderstanding `ActiveSupport::Dependencies` in Thinking Sphinx? Or something else entirely? :)

**Comment by casperisfine** [comment]
`unloadable_cpaths` is not an actual property, it's just a shortcut, see it's implementation: https://github.com/fxn/zeitwerk/blob/6b824c2ee5342134a9e1dc518fb961bd110d1b20/lib/zeitwerk/loader.rb#L409-L411

**Comment by pat** [comment]
@casperisfine yup - and the result there is frozen. I'm curious as to why it's frozen - given the result is used in Dependencies methods where modifications can take place - because I'm sure there's a good reason for it, but I don't understand the broader implications.

I could just go and overload that method, but I'd prefer to know what I'm messing with :)

**Comment by casperisfine** [comment]
> I'm curious as to why it's frozen

To indicate that mutating it would have no effect.

> the result is used in Dependencies methods where modifications can take place

I don't think it's accurate. I believe it's there for reading only.

**Comment by casperisfine** [comment]
Ok, I think I now see what you are getting at:

```ruby
    # An array of qualified constant names that have been loaded. Adding a name
    # to this array will cause it to be unloaded the next time Dependencies are
    # cleared.
    mattr_accessor :autoloaded_constants, default: []
```

I can't tell for Xavier if it's accidental or on purpose, but this feature is no longer supported with Zeitwerk.

**Comment by pat** [comment]
I think you've picked up what I'm attempting to say, but for anyone else looking at this issue, here's a bit more of what I'm getting at:

The value of `autoloaded_constants` can be modified:
https://github.com/rails/rails/blob/master/activesupport/lib/active_support/dependencies.rb#L490-L491

… and while it is initially defined as a module accessor with a default value of an empty array:
https://github.com/rails/rails/blob/master/activesupport/lib/active_support/dependencies.rb#L81

… it is overwritten as part of the Zeitwerk integration:
https://github.com/rails/rails/blob/master/activesupport/lib/active_support/dependencies/zeitwerk_integration.rb#L26-L28
https://github.com/rails/rails/blob/master/activesupport/lib/active_support/dependencies/zeitwerk_integration.rb#L92

And the value obtained through that integration is frozen, which doesn't seem reasonable given the modifications that can occur:
https://github.com/fxn/zeitwerk/blob/6b824c2ee5342134a9e1dc518fb961bd110d1b20/lib/zeitwerk/loader.rb#L409-L411

**Comment by fxn** [comment]
Hi Pat! Long time!

The integration walks a fine line because most of the API is undocumented, and most of it won't be ported. In particular, `require_or_load` is gone in `zeitwerk` mode, because this mode does not have a "mechanism" anymore. I need to find a balance in what is supported, and from what is not supported, what has to be documented. I believe this aspect of the integration can be improved and I keep an open mind to revise details, while at the same time defaulting on being conservative. This is an opportunity to trim down all those methods accumulated over the years.

The API that is supported is basically the one you can see in [this file](https://github.com/rails/rails/blob/master/activesupport/lib/active_support/dependencies/zeitwerk_integration.rb). I pondered whether to implement everything else raising, but by know it is not done that way, unless actual real usage proves raising is better in practice.

`require_dependency` is a method that is meant to work in both autoloading modes, would that work in your use case?

**Comment by fxn** [comment]
Let me also link to this [summary](https://gist.github.com/fxn/271deed5a0772a5967c3aa00a8306746) of how the integration is implemented, in case it gives more context.

**Comment by pat** [comment]
Thanks Xavier for those details - and hi, it has been a while! :)

I've just tried `require_dependency` - but it expects a constant to be defined in the given file that matches the file name. That behaviour works in many cases, but sadly not when Thinking Sphinx loads indices, and so a `NameError` is raised. `require` hits the same problem as well (when called within an object's context - Rails has replaced the Ruby approach, I'm guessing).

My reason for using `ActiveSupport::Dependencies.require_or_load` in the first place is because I want the index files (and the structures within them) to be reloaded in a development environment, like any model/controller/etc. I've taken this approach since Rails 3.0, when I rewrote Thinking Sphinx 8 years ago (😱).

I'm now testing it with `load`, and the index details are not (incorrectly) cached between requests on a development Rails server, so I think that confirms it provides the behaviour I'm looking for. If there are any suggestions on a better approach for this, do let me know! But given my initial question, I'm going to close this issue. Thanks again to both of you for the responses.
